File: intellinum/intellinum/report/new_leave_allocation_report/new_leave_allocation_report.py
# # Copyright (c) 2026, ibsl and contributors
# # For license information, please see license.txt
#
# # import frappe
#
#
# # def execute(filters=None):
# # 	columns, data = [], []
# # 	return columns, data



# Copyright (c) 2026, ibsl and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {"label": _("Employee"),                        "fieldname": "employee",                        "fieldtype": "Link", "options": "Employee", "width": 100},
        {"label": _("Employee Name"),                   "fieldname": "employee_name",                   "fieldtype": "Data",  "width": 220},
        {"label": _("Employment Type"),                 "fieldname": "employment_type",                 "fieldtype": "Data",  "width": 130},
        {"label": _("Date of Joining"),                 "fieldname": "date_of_joining",                 "fieldtype": "Date",  "width": 120},
        {"label": _("Confirmation Date"),               "fieldname": "final_confirmation_date",         "fieldtype": "Date",  "width": 140},
        {"label": _("Annual Leave"),                    "fieldname": "annual_leave",                    "fieldtype": "Float", "width": 110},
        {"label": _("Leave to be Allocated"),           "fieldname": "leave_to_be_allocated",           "fieldtype": "Float", "width": 160},
        {"label": _("Optional Holidays"),               "fieldname": "optional_holidays",               "fieldtype": "Float", "width": 130},
        {"label": _("Optional Holidays to Allocate"),   "fieldname": "optional_holidays_to_allocate",   "fieldtype": "Float", "width": 190},
        {"label": _("Status"),                          "fieldname": "status",                          "fieldtype": "Data",  "width": 110},
        {"label": _("Action"),                          "fieldname": "action",                          "fieldtype": "Data",  "width": 130},
    ]

def get_data(filters):
    conditions = "WHERE e.final_confirmation_date IS NOT NULL AND e.status = 'Active'"
    if filters:
        if filters.get("from_date"):
            conditions += " AND e.final_confirmation_date >= %(from_date)s"
        if filters.get("to_date"):
            conditions += " AND e.final_confirmation_date <= %(to_date)s"

    data = frappe.db.sql(
        """
        SELECT
            e.name AS employee,
            e.employee_name,
            e.employment_type,
            e.date_of_joining,
            e.final_confirmation_date,
            21.0 AS annual_leave,
            LEAST(
                ROUND(
                    (DATEDIFF(CONCAT(YEAR(CURDATE()), '-12-31'), e.final_confirmation_date) + 1)
                    / 365.0 * 21,
                    2
                ),
                21.0
            ) AS leave_to_be_allocated,
            4.0 AS optional_holidays,
            LEAST(
                ROUND(
                    (DATEDIFF(
                        CONCAT(YEAR(CURDATE()), '-12-31'),
                        CASE
                            WHEN YEAR(e.date_of_joining) = YEAR(CURDATE())
                                THEN e.date_of_joining
                            ELSE CONCAT(YEAR(CURDATE()), '-01-01')
                        END
                    ) + 1)
                    / 365.0 * 4,
                    2
                ),
                4.0
            ) AS optional_holidays_to_allocate,
            CASE
                WHEN la.name IS NOT NULL THEN 'Allocated'
                ELSE 'Pending'
            END AS status,
            la.name AS allocation_name
        FROM `tabEmployee` e
        LEFT JOIN `tabLeave Allocation` la
            ON  la.employee   = e.name
            AND la.leave_type = 'Paid Leave'
            AND la.docstatus  = 1
            AND YEAR(la.to_date) = YEAR(CURDATE())
        {conditions}
        ORDER BY e.name
        """.format(conditions=conditions),
        filters or {},
        as_dict=True,
    )

    for row in data:
        row["action"] = "Allocate" if row.get("status") == "Pending" else ""

    return data

File: intellinum/intellinum/report/new_leave_allocation_report/new_leave_allocation_report.js
// Copyright (c) 2026, ibsl and contributors
// For license information, please see license.txt

// frappe.query_reports["New Leave Allocation Report"] = {
// 	"filters": [
//
// 	]
// };








frappe.query_reports["New Leave Allocation Report"] = {

    filters: [
        { fieldname: "from_date", label: __("From Date"), fieldtype: "Date" },
        { fieldname: "to_date",   label: __("To Date"),   fieldtype: "Date" },
    ],

    formatter: function (value, row, column, data, default_formatter) {

        if (column.fieldname === "action") {
            if (data && data.status === "Pending") {
                const emp     = (data.employee || "").replace(/'/g, "\\'");
                const empName = (data.employee_name || "").replace(/'/g, "\\'");
                const conf    = (data.final_confirmation_date || "");
                const leaves  = parseInt(data.leave_to_be_allocated) || 0;

                return `<button
                    class="btn btn-default btn-xs jb-alloc-btn"
                    onclick="window._jbAllocate(this,'${emp}','${empName}','${conf}',${leaves})">
                    Allocate
                </button>`;
            }
            return data && data.status === "Allocated" ? `&#10003; Done` : "";
        }

        return default_formatter(value, row, column, data);
    },

    onload: function (report) {

        window._jbAllocate = function (btn, employee, empName, confDate, leaves) {
            const yearEnd = frappe.datetime.year_end();

            frappe.confirm(
                `Do you want to allocate ${leaves} Paid Leave(s) to ${empName}?<br>
                 <small>From: ${confDate} &nbsp;&rarr;&nbsp; To: ${yearEnd}</small>`,

                function () {
                    const _origMsgprint = frappe.msgprint;
                    frappe.msgprint = () => {};

                    // Disable button immediately to prevent double click
                    $(btn).prop("disabled", true).text("Allocating...");

                    frappe.call({
                        method: "frappe.client.insert",
                        args: {
                            doc: {
                                doctype             : "Leave Allocation",
                                employee            : employee,
                                leave_type          : "Paid Leave",
                                from_date           : confDate,
                                to_date             : yearEnd,
                                new_leaves_allocated: leaves,
                            },
                        },
                        freeze        : true,
                        freeze_message: "Creating Leave Allocation...",
                        error_handlers: {
                            OverlapError: function() {
                                frappe.msgprint = _origMsgprint;
                                // Already allocated — update row instantly
                                window._jbUpdateRow(btn);
                                frappe.show_alert({ message: `Already allocated for ${empName}.`, indicator: "orange" });
                            },
                            OverAllocationError: function() {
                                frappe.msgprint = _origMsgprint;
                                $(btn).prop("disabled", false).text("Allocate");
                                frappe.show_alert({ message: `Exceeds max limit for ${empName}.`, indicator: "red" });
                            }
                        },
                        callback: function (r) {
                            if (r.exc || !r.message) {
                                frappe.msgprint = _origMsgprint;
                                $(btn).prop("disabled", false).text("Allocate");
                                return;
                            }
                            frappe.call({
                                method: "frappe.client.submit",
                                args  : { doc: r.message },
                                freeze: true,
                                freeze_message: "Submitting...",
                                callback: function (res) {
                                    frappe.msgprint = _origMsgprint;
                                    if (res.exc) {
                                        $(btn).prop("disabled", false).text("Allocate");
                                        frappe.show_alert({ message: "Submit failed.", indicator: "red" });
                                        return;
                                    }
                                    // ✅ Instantly update row — no page refresh
                                    window._jbUpdateRow(btn);
                                    frappe.show_alert({
                                        message  : `Leave allocated for ${empName}!`,
                                        indicator: "green",
                                    });
                                },
                            });
                        },
                    });
                },
                // On "No" — re-enable button
                function() {
                    $(btn).prop("disabled", false).text("Allocate");
                }
            );
        };

        // Updates Status cell → "Allocated", Action cell → "✓ Done" instantly
        window._jbUpdateRow = function(btn) {
            const $cell   = $(btn).closest(".dt-cell");
            const $row    = $cell.closest(".dt-row");

            // Update Action cell
            $cell.find(".dt-cell__content").html("&#10003; Done");

            // Find Status cell in same row and update it
            $row.find(".dt-cell").each(function() {
                const colId = $(this).attr("data-col-index");
                const text  = $(this).find(".dt-cell__content").text().trim();
                if (text === "Pending") {
                    $(this).find(".dt-cell__content").text("Allocated");
                }
            });
        };
    },
};
